### ISSUE TITLE
App crashing when using LightenedCNN_A model

### ISSUE THREAD

**Issue by PriyaSindkar** [MeloniZippoProjects/MultimediaProject#2]
I want to have a light model (light in terms of size in MB) so I switched the model from resnet50 to LightenedCNN_A model. I had to change the forward() method called from extract() method of DNNExtractor class from 

`net.forward(outputBlobs, Parameters.DEEP_LAYER);`
to 
`net.forward(outputBlobs);`

But I am facing 2 problems with these changes-

1. I am getting "No faces detected" although the white box around face is showing accurately. Thus, nothing is being saved under "unspecified". 
2. When I add an identity from gallery or capture from video and when I click on "Save Identity" button, the app crashes.

Here is the stacktrace for the crash-

`E/cv::error(): OpenCV(3.4.1) Error: Assertion failed (srcMat.dims == 2 && srcMat.cols == weights.cols && dstMat.rows == srcMat.rows && dstMat.cols == weights.rows && srcMat.type() == weights.type() && weights.type() == dstMat.type() && srcMat.type() == 5 && (biasMat.empty() || (biasMat.type() == srcMat.type() && biasMat.isContinuous() && (int)biasMat.total() == dstMat.cols))) in static void cv::dnn::FullyConnectedLayerImpl::FullyConnected::run(const cv::Mat&, const cv::Mat&, const cv::Mat&, cv::Mat&, const cv::dnn::experimental_dnn_v4::ActivationLayer*, int), file /build/master_pack-android/opencv/modules/dnn/src/layers/fully_connected_layer.cpp, line 154
2019-08-14 18:10:39.017 24057-24174/melonizippo.org.facerecognition E/org.opencv.dnn: dnn::forward_13() caught cv::Exception: OpenCV(3.4.1) /build/master_pack-android/opencv/modules/dnn/src/layers/fully_connected_layer.cpp:154: error: (-215) srcMat.dims == 2 && srcMat.cols == weights.cols && dstMat.rows == srcMat.rows && dstMat.cols == weights.rows && srcMat.type() == weights.type() && weights.type() == dstMat.type() && srcMat.type() == 5 && (biasMat.empty() || (biasMat.type() == srcMat.type() && biasMat.isContinuous() && (int)biasMat.total() == dstMat.cols)) in function static void cv::dnn::FullyConnectedLayerImpl::FullyConnected::run(const cv::Mat&, const cv::Mat&, const cv::Mat&, cv::Mat&, const cv::dnn::experimental_dnn_v4::ActivationLayer*, int)
2019-08-14 18:10:39.019 24057-24174/melonizippo.org.facerecognition E/AndroidRuntime: FATAL EXCEPTION: Thread-4
    Process: melonizippo.org.facerecognition, PID: 24057
    CvException [org.opencv.core.CvException: cv::Exception: OpenCV(3.4.1) /build/master_pack-android/opencv/modules/dnn/src/layers/fully_connected_layer.cpp:154: error: (-215) srcMat.dims == 2 && srcMat.cols == weights.cols && dstMat.rows == srcMat.rows && dstMat.cols == weights.rows && srcMat.type() == weights.type() && weights.type() == dstMat.type() && srcMat.type() == 5 && (biasMat.empty() || (biasMat.type() == srcMat.type() && biasMat.isContinuous() && (int)biasMat.total() == dstMat.cols)) in function static void cv::dnn::FullyConnectedLayerImpl::FullyConnected::run(const cv::Mat&, const cv::Mat&, const cv::Mat&, cv::Mat&, const cv::dnn::experimental_dnn_v4::ActivationLayer*, int)
    ]
        at org.opencv.dnn.Net.forward_3(Native Method)
        at org.opencv.dnn.Net.forward(Net.java:302)
        at melonizippo.org.facerecognition.deep.DNNExtractor.extract(DNNExtractor.java:40)
        at melonizippo.org.facerecognition.AddIdentityActivity.addImage(AddIdentityActivity.java:626)
        at melonizippo.org.facerecognition.AddIdentityActivity.addImage(AddIdentityActivity.java:605)
        at melonizippo.org.facerecognition.AddIdentityActivity.addImage(AddIdentityActivity.java:598)
        at melonizippo.org.facerecognition.AddIdentityActivity.lambda$processImages$10$AddIdentityActivity(AddIdentityActivity.java:533)
        at melonizippo.org.facerecognition.AddIdentityActivity$$Lambda$8.run(Unknown Source:20)
        at java.lang.Thread.run(Thread.java:764)`

I am missing something? Some change I am failing to do to make this model work for face detection and saving the identity?

**Comment by enricomeloni** [comment]
Hi @PriyaSindkar,

The instruction `net.forward(outputBlobs, Parameters.DEEP_LAYER);` stops the processing of the network at the layer with the name contained in `Parameters.DEEP_LAYER` and returns the output of that layer. 

That layer (`pool5/7x7_s1`) is  the output of the final Convolutional Layer in resnet50, just before the Fully Connected Layer which would recognize actors' faces. 

The output of this layer can be used as a feature vector, which is what we do.

LightenedCNN has a different structure than resnet50, therefore you should adapt the code:

- the feature vector could be of different size, resnet is 1x2048, lightenedCNN could be something else.
- the last convolutional layer could have a different name (change `Parameters.DEEP_LAYER` to that one)

After that, faces should be found and identities saved.

**Comment by PriyaSindkar** [comment]
@enricomeloni Thank you for detailed explanation. I am using "fc2" as the output layer for lightenedCNN. But still getting same result. Can you guide me how to get the correct feature vector size?

EDIT
I changed the output layer to "softmax" and I made the images to gray scale with input sizes as 128x128. Now I am able to get identify faces but now the problem is that when I use 2 faces for testing, it is given me wrong face recognition.

**Comment by rzippo** [comment]
I am not sure I understand the problem you have after the last edit. Were you able to find the correct input size and feature vector size? Were you able to identify the layer to use as the feature vector? Is the recognition problem a bug or a performance issue?

**Comment by PriyaSindkar** [comment]
Yes, I found out the correct layer, input size. Problem is- when I test the app with 3 different faces with approx. 14 training data for 2 faces and 30-50 training data for 1 face, the app is showing incorrect recognition (5 out of 5 times approximately) for the first 2 faces which are with less training data. It is a recognition problem.
